CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+
+### cardano_check_read_access(): Promise\<bool>
+
+Errors: APIError
+
+Returns true if the full API (`cardano` object) was injected and the dApp has access to the user's wallet, and false otherwise.
+
+
+
+
+## Full API
+
+Upon successful connection via `cardano_request_read_access()`, a javascript object named `cardano` is injected with the following methods. All read-only methods (all but the signing functionality) should not require any user interaction as the user has already consented to the dApp reading information about the wallet's state when they agreed to `cardano_request_read_access()`. The remaining methods `cardano.sign_tx()`, `cardano.sign_tx_input()`, `cardano.sign_data()` must request the user's consent in an informative way for each and every API call in order to maintain security.
+
+The API chosen here is for the minimum API necessary for dApp <-> Wallet interactions without convenience functions that don't strictly need the wallet's state to work. The API here is for now also only designed for Shelley's Mary hardfork and thus has NFT support. When Alonzo is released with Plutus support this API will have to be extended.
+
+### cardano.get_utxos(amount: cbor\<value> = undefined, paginate: Paginate = undefined): Promise\<TransactionUnspentOutput[] | undefined>
+
+Errors: `APIError`, `PaginateError`
+
+If `amount` is `undefined`, this shall return a list of all UTXOs (unspent transaction outputs) controlled by the wallet. If `amount` is not `undefined`, this request shall be limited to just the UTXOs that are required to reach the combined ADA/multiasset value target specified in `amount`, and if this cannot be attained, `undefined` shall be returned. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_balance(): Promise\<cbor\<value>>
+
+Errors: `APIError`
+
+Returns the total balance available of the wallet. This is the same as summing the results of `cardano.get_utxos()`, but it is both useful to dApps and likely already maintained by the implementing wallet in a more efficient manner so it has been included in the API as well.
+
+### cardano.get_used_addresses(paginate: Paginate = undefined): Promise\<cbor\<address>[]>
+
+Errors: `APIError`
+
+Returns a list of all used (included in some on-chain transaction) addresses controlled by the wallet. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_unused_addresses(): Promise\<cbor\<address>[]>
+
+Errors: `APIError`
+
+Returns a list of unused addresses controlled by the wallet.
+
+### cardano.get_change_address(): Promise\<cbor\<address>>
```

Comment:
I feel this API is not flexible enough to support any dApp that wants to use anything other than base addresses. Single the API only allows for a single return address, presumably the wallet returns just a single base address? What if the dApp wants instead a pointer address or an enterprise address?

Beyond just the address type problem, the fact it returns a single address may be problematic if the dApp wants to interact with enterprise addresses for some privacy reason. Specifically, although input selection for Yoroi uses a single address for change, cardano-wallet uses multiple change addresses.
The reason multiple change addresses existed is because there is a limit to how big a single UTXO entry can be so the wallet may need multiple UTXOs to give back all the change. If this was all given to the same address, it would make it easy to know which output address belongs to the user.

Maybe all of this is over-complicating things though

Reply:
>For this kind of situation where you can't inject code directly, people in Ethereum usually use another tool called Wallet Connect which has a different flow for setting up the connection

From what I could tell, this is only for mobile wallets?

Reply:
It's used on desktop as well. You can see in this tweet for example the flow of picking which connector they want to use, then selecting Wallet Connect and using their desktop app to complete the connection
https://twitter.com/CeloTerminal/status/1400079861931249667

Reply:
For the single address, would it be sufficient to just switch it to return an array of addresses, which the dApp could then use as many as it needs? Or do we need to make it take some kind of parameter (perhaps defaulting to single?) to specify which kind of address you want? Although the dApp would be able to easily construct an enterprise address from a single address, wouldn't it?